### ISSUE TITLE
Expression Optimization Error

### ISSUE THREAD

**Issue by billdenney** [nlmixrdevelopment/RxODE#280]
I thought that this example may help make the expression optimization code more robust.  The use of `foceiControl(optExpression=FALSE)` fixed my real example.

``` r
library(nlmixr)
library(tidyverse)
nlmixr_model_pd <- function(){
  ini({
    tvKA <- fixed(log(0.5))    ; label("Drug absorption rate (1/hr)")
    tvCL <- fixed(log(200))  ; label("Drug clearance (mL/hr)")
    tvVc <- fixed(log(100))  ; label("Drug central volume (mL)")
    
    tvLIGAND <- log(100)
    tvKK <- log(0.001)
    tvKOUT <- log(0.0001)
    
    #eta_LIGAND <- fixed(0)
    eta_LIGAND ~ 0.1
    eta_KK <- fixed(0)
    #eta_KK ~ 0.1
    
    err_add <- 0.1
    err_prop <- 0.1
  })
  
  model({
    LIGAND <- exp(tvLIGAND+eta_LIGAND)
    KK <- exp(tvKK+eta_KK)
    KOUT <- exp(tvKOUT)
    
    # Or if the same animals/patients had PK and PD sampling, then instead of
    # population PK estimates use the individual PK parameter estimates by
    # inserting their individual values:
    KA<-exp(tvKA)*(WTKG/0.02)^-0.25
    CL<-exp(tvCL)*(WTKG/0.02)^0.75
    VC<-exp(tvVc)*(WTKG/0.02)
    
    K20 = CL/VC
    S2 = VC
    
    KIN = KOUT*LIGAND
    
    d/dt(A1) = -KA*A1
    d/dt(A2) =  KA*A1 - K20*A2
    
    CP = A2/S2
    
    A3(0) = LIGAND
    d/dt(A3) = KIN - KOUT*A3 - KK*CP*A3
    
    IFN = A3
    IFN ~ add(err_add) + prop(err_prop)
  })
}

fit_pd <- nlmixr(nlmixr_model_pd, theo_sd %>% mutate(WTKG=0.02), est="focei")
#> i parameter labels from comments will be replaced by 'label()'
#> > creating full model...
#> > pruning branches (`if`/`else`)...
#> v done
#> > loading into symengine environment...
#> v done
#> > creating full model...
#> > pruning branches (`if`/`else`)...
#> v done
#> > loading into symengine environment...
#> v done
#> > calculate jacobian
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> > calculate sensitivities
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> > calculate d(f)/d(eta)
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> > calculate d(R^2)/d(eta)
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> > finding duplicate expressions in inner model...
#> Error in ..rxOpt(x): error optimizing expression, try 'optExpression=FALSE'
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> Error: stopped optimizing duplicate expressions
#> Error: Aborted calculation
#fit_pd <- nlmixr(nlmixr_model_pd, theo_sd %>% mutate(WTKG=0.02), est="focei", control=foceiControl(optExpression=FALSE))
```

<sup>Created on 2020-09-07 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by mattfidler** [comment]
Thanks Bill

**Comment by mattfidler** [comment]
Smaller reprex:

``` r
library(RxODE)
#> RxODE 0.9.2.0 using 4 threads (see ?getRxThreads)
rxOptExpr("
    cmt(A1)
    cmt(A2)
    cmt(A3)
    d/dt(A1) = -0.376060309308639 * exp(THETA[1]) * A1 * WTKG^(-0.25)
    d/dt(A2) = 0.376060309308639 * exp(THETA[1]) * A1 * WTKG^(-0.25) - 
        0.376060309308639 * exp(THETA[2] - THETA[3]) * A2 * WTKG^(-0.25)
    A3(0) = exp(ETA[1] + THETA[4])
    d/dt(A3) = -exp(THETA[6]) * A3 - 0.02 * exp(-THETA[3] + THETA[5] + 
        THETA[7]) * A3 * A2/WTKG + exp(ETA[1] + THETA[4] + THETA[6])
    d/dt(rx__sens_A1_BY_ETA_1___) = -0.376060309308639 * exp(THETA[1]) * 
        WTKG^(-0.25) * rx__sens_A1_BY_ETA_1___
    d/dt(rx__sens_A2_BY_ETA_1___) = 0.376060309308639 * exp(THETA[1]) * 
        WTKG^(-0.25) * rx__sens_A1_BY_ETA_1___ - 0.376060309308639 * 
        exp(THETA[2] - THETA[3]) * WTKG^(-0.25) * rx__sens_A2_BY_ETA_1___
    d/dt(rx__sens_A3_BY_ETA_1___) = (-0.02 * exp(-THETA[3] + 
        THETA[5] + THETA[7]) * A2/WTKG - exp(THETA[6])) * rx__sens_A3_BY_ETA_1___ - 
        0.02 * exp(-THETA[3] + THETA[5] + THETA[7]) * A3 * rx__sens_A2_BY_ETA_1___/WTKG + 
        exp(ETA[1] + THETA[4] + THETA[6])
    rx__sens_A3_BY_ETA_1___(0) = exp(ETA[1] + THETA[4]) + 0
    rx_yj_ ~ 2
    rx_lambda_ ~ 1
    rx_hi_ ~ 1
    rx_low_ ~ 0
    rx_pred_ = A3
    rx__sens_rx_pred__BY_ETA_1___ = rx__sens_A3_BY_ETA_1___
    rx_r_ = (THETA[8] + abs(A3) * THETA[9])^2
    rx__sens_rx_r__BY_ETA_1___ = 2 * (THETA[8] + abs(A3) * THETA[9]) * 
        dabs(A3) * THETA[9] * rx__sens_A3_BY_ETA_1___
    cmt(IFN)
")
#> → finding duplicate expressions in model...
#> Error in ..rxOpt(x): error optimizing expression, try 'optExpression=FALSE'
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> Error: stopped optimizing duplicate expressions
```

<sup>Created on 2020-09-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by mattfidler** [comment]
Fixed

<details>

``` r
library(RxODE)
#> RxODE 0.9.2.0 using 4 threads (see ?getRxThreads)
rxOptExpr("
    cmt(A1)
    cmt(A2)
    cmt(A3)
    d/dt(A1) = -0.376060309308639 * exp(THETA[1]) * A1 * WTKG^(-0.25)
    d/dt(A2) = 0.376060309308639 * exp(THETA[1]) * A1 * WTKG^(-0.25) - 
        0.376060309308639 * exp(THETA[2] - THETA[3]) * A2 * WTKG^(-0.25)
    A3(0) = exp(ETA[1] + THETA[4])
    d/dt(A3) = -exp(THETA[6]) * A3 - 0.02 * exp(-THETA[3] + THETA[5] + 
        THETA[7]) * A3 * A2/WTKG + exp(ETA[1] + THETA[4] + THETA[6])
    d/dt(rx__sens_A1_BY_ETA_1___) = -0.376060309308639 * exp(THETA[1]) * 
        WTKG^(-0.25) * rx__sens_A1_BY_ETA_1___
    d/dt(rx__sens_A2_BY_ETA_1___) = 0.376060309308639 * exp(THETA[1]) * 
        WTKG^(-0.25) * rx__sens_A1_BY_ETA_1___ - 0.376060309308639 * 
        exp(THETA[2] - THETA[3]) * WTKG^(-0.25) * rx__sens_A2_BY_ETA_1___
    d/dt(rx__sens_A3_BY_ETA_1___) = (-0.02 * exp(-THETA[3] + 
        THETA[5] + THETA[7]) * A2/WTKG - exp(THETA[6])) * rx__sens_A3_BY_ETA_1___ - 
        0.02 * exp(-THETA[3] + THETA[5] + THETA[7]) * A3 * rx__sens_A2_BY_ETA_1___/WTKG + 
        exp(ETA[1] + THETA[4] + THETA[6])
    rx__sens_A3_BY_ETA_1___(0) = exp(ETA[1] + THETA[4]) + 0
    rx_yj_ ~ 2
    rx_lambda_ ~ 1
    rx_hi_ ~ 1
    rx_low_ ~ 0
    rx_pred_ = A3
    rx__sens_rx_pred__BY_ETA_1___ = rx__sens_A3_BY_ETA_1___
    rx_r_ = (THETA[8] + abs(A3) * THETA[9])^2
    rx__sens_rx_r__BY_ETA_1___ = 2 * (THETA[8] + abs(A3) * THETA[9]) * 
        dabs(A3) * THETA[9] * rx__sens_A3_BY_ETA_1___
    cmt(IFN)
")
#> → finding duplicate expressions in model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → optimizing duplicate expressions in model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> [1] "cmt(A1)\ncmt(A2)\ncmt(A3)\nrx_expr_1~WTKG^(-0.25)\nrx_expr_2~exp(THETA[1])\nd/dt(A1)=-0.376060309308639*rx_expr_2*A1*rx_expr_1\nrx_expr_6~THETA[2]-THETA[3]\nrx_expr_8~exp(rx_expr_6)\nrx_expr_12~0.376060309308639*rx_expr_2\nrx_expr_15~0.376060309308639*rx_expr_8\nd/dt(A2)=rx_expr_12*A1*rx_expr_1-rx_expr_15*A2*rx_expr_1\nrx_expr_4~ETA[1]+THETA[4]\nrx_expr_7~exp(rx_expr_4)\nA3(0)=rx_expr_7\nrx_expr_3~exp(THETA[6])\nrx_expr_9~rx_expr_4+THETA[6]\nrx_expr_11~exp(rx_expr_9)\nrx_expr_13~exp(-THETA[3]+THETA[5]+THETA[7])\nrx_expr_14~0.02*rx_expr_13\nrx_expr_16~rx_expr_14*A3\nd/dt(A3)=-rx_expr_3*A3-rx_expr_16*A2/WTKG+rx_expr_11\nd/dt(rx__sens_A1_BY_ETA_1___)=-0.376060309308639*rx_expr_2*rx_expr_1*rx__sens_A1_BY_ETA_1___\nd/dt(rx__sens_A2_BY_ETA_1___)=rx_expr_12*rx_expr_1*rx__sens_A1_BY_ETA_1___-rx_expr_15*rx_expr_1*rx__sens_A2_BY_ETA_1___\nd/dt(rx__sens_A3_BY_ETA_1___)=(-0.02*rx_expr_13*A2/WTKG-rx_expr_3)*rx__sens_A3_BY_ETA_1___-rx_expr_16*rx__sens_A2_BY_ETA_1___/WTKG+rx_expr_11\nrx__sens_A3_BY_ETA_1___(0)=rx_expr_7\nrx_yj_~2\nrx_lambda_~1\nrx_hi_~1\nrx_low_~0\nrx_pred_=A3\nrx__sens_rx_pred__BY_ETA_1___=rx__sens_A3_BY_ETA_1___\nrx_expr_0~abs(A3)\nrx_expr_5~rx_expr_0*THETA[9]\nrx_expr_10~THETA[8]+rx_expr_5\nrx_r_=((rx_expr_10)*(rx_expr_10))\nrx__sens_rx_r__BY_ETA_1___=2*(rx_expr_10)*dabs(A3)*THETA[9]*rx__sens_A3_BY_ETA_1___\ncmt(IFN)"
```

<sup>Created on 2020-09-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>


``` r
library(nlmixr)
library(tidyverse)
nlmixr_model_pd <- function(){
  ini({
    tvKA <- fixed(log(0.5))    ; label("Drug absorption rate (1/hr)")
    tvCL <- fixed(log(200))  ; label("Drug clearance (mL/hr)")
    tvVc <- fixed(log(100))  ; label("Drug central volume (mL)")
    
    tvLIGAND <- log(100)
    tvKK <- log(0.001)
    tvKOUT <- log(0.0001)
    
    #eta_LIGAND <- fixed(0)
    eta_LIGAND ~ 0.1
    eta_KK <- fixed(0)
    #eta_KK ~ 0.1
    
    err_add <- 0.1
    err_prop <- 0.1
  })
  
  model({
    LIGAND <- exp(tvLIGAND+eta_LIGAND)
    KK <- exp(tvKK+eta_KK)
    KOUT <- exp(tvKOUT)
    
    # Or if the same animals/patients had PK and PD sampling, then instead of
    # population PK estimates use the individual PK parameter estimates by
    # inserting their individual values:
    KA<-exp(tvKA)*(WTKG/0.02)^-0.25
    CL<-exp(tvCL)*(WTKG/0.02)^0.75
    VC<-exp(tvVc)*(WTKG/0.02)
    
    K20 = CL/VC
    S2 = VC
    
    KIN = KOUT*LIGAND
    
    d/dt(A1) = -KA*A1
    d/dt(A2) =  KA*A1 - K20*A2
    
    CP = A2/S2
    
    A3(0) = LIGAND
    d/dt(A3) = KIN - KOUT*A3 - KK*CP*A3
    
    IFN = A3
    IFN ~ add(err_add) + prop(err_prop)
  })
}

fit_pd <- nlmixr(nlmixr_model_pd, theo_sd %>% mutate(WTKG=0.02), est="focei",control=foceiControl(print=0))
#> ℹ parameter labels from comments will be replaced by 'label()'
#> → creating full model...
#> → pruning branches (`if`/`else`)...
#> ✔ done
#> → loading into symengine environment...
#> ✔ done
#> → creating full model...
#> → pruning branches (`if`/`else`)...
#> ✔ done
#> → loading into symengine environment...
#> ✔ done
#> → calculate jacobian
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → calculate sensitivities
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → calculate ∂(f)/∂(η)
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → calculate ∂(R²)/∂(η)
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → finding duplicate expressions in inner model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → optimizing duplicate expressions in inner model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → finding duplicate expressions in EBE model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → optimizing duplicate expressions in EBE model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → compiling inner model...
#> ✔ done
#> → compiling EBE model...
#> ✔ done
#> Needed Covariates:
#> [1] "WTKG"
#> RxODE 0.9.2.0 using 4 threads (see ?getRxThreads)
#> Theta reset
#> [====|====|====|====|====|====|====|====|====|====] 0:00:04 
#> done
#> Calculating residuals/tables
#> done.
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod =
#> FALSE, : initial ETAs were nudged; (can control by foceiControl(etaNudge=.))
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod
#> = FALSE, : ETAs were reset to zero during optimization; (Can control by
#> foceiControl(resetEtaP=.))
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod = FALSE, : parameter estimate near boundary; covariance not calculated
#>  use 'getVarCov' to calculate anyway
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod =
#> FALSE, : gradient problems with initial estimate; see $scaleInfo
```

<sup>Created on 2020-09-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

</details>

**Comment by billdenney** [comment]
Thanks!
